### ISSUE TITLE
OSX Safari 11.1 room leave causes crash

### ISSUE THREAD

**Issue by wesleykendall** [twilio/twilio-video.js#294]
- [x] I have verified that the issue occurs with the latest twilio-video.js release and is not marked as a known issue in the [CHANGELOG.md](https://github.com/twilio/twilio-video.js/blob/master/CHANGELOG.md).
 - [x] I reviewed the [Common Issues](https://github.com/twilio/twilio-video.js/blob/master/COMMON_ISSUES.md) and open GitHub issues and verified that this report represents a potentially new issue.
 - [x] I verified that the [Quickstart](https://github.com/twilio/video-quickstart-js) application works in my environment (this application causes the crash)

**Code to reproduce the issue:**

Open https://github.com/twilio/video-quickstart-js on the latest Safari Version 11.1 (13605.1.33.1.4) and join a room. Then open the quickstart on any other browser (including safari on another computer over ngrok). Click "leave room" on any of the browsers and Safari will refresh the page with a message that says "This webpage was reloaded because a problem occurred"

**Expected behavior:**

The expected behavior for leaving a room is for it to disconnect both participants and remove the tracks successfully. This behavior happens on Chrome and Firefox

**Actual behavior:**

On Safari, the browser abruptly refreshes with "This webpage was reloaded because a problem occurred". It also clears the Javascript console and does not allow me to see any logs

**Software versions:**

 - [X] Browser(s): OSX Safari Version 11.1 (13605.1.33.1.4)
 - [X] Operating System: High Sierra 10.13.4 (17E202)
 - [X] twilio-video.js: v1.8.0 and also tested it with 1.9.0-dev

**Additional info:**

I've been developing a video application and have had issues exclusively with Safari on OSX (and ios). I've tried peer to peer rooms and group rooms supporting all codecs but still get the same behavior of applications crashing when participants leave rooms.

I noticed that https://github.com/twilio/video-quickstart-js also has the same behavior. When a participant leaves a room, Safari refreshes and restarts. I've verified this on two different Mac machines and on an iphone.

If you have any idea of what might be going on, I'd greatly appreciate it. I've tried using the twilio-video 1.9 dev version as well. Unfortunately I've found no way to prevent Safari from refreshing after this particular crash, which has made it even more frustrating to find logs or errors related to this problem. Thanks for the help!

**Comment by markandrus** [comment]
Hi @wesleykendall,

Sorry you're hitting this. I filed [Bug 185228 - EXC_BAD_ACCESS using WebRTC and MediaStream APIs](https://bugs.webkit.org/show_bug.cgi?id=185228) describing this same issue. In the bug tracker, they said [Bug 185233 - A MediaStream being played should allow removing some of its tracks](https://bugs.webkit.org/show_bug.cgi?id=185233) should fix this. I'm not sure when it will be released, though.

Mark

**Comment by wesleykendall** [comment]
Thanks @markandrus ! Do you happen to now if there is a hack to get around it programmatically? For example, refreshing the browser when a client disconnects?

**Comment by ibrainventures** [comment]
Hi there,

we switched today from simple-webrtc to twilio-video.js and got the same issue. Ios 11.3.1 and after exit from the remote-client, the ios-safari crashes and make a auto-reload ... (we are wrapping the twilio inside vue / quasar).

**Comment by wesleykendall** [comment]
@ibrainventures we are looking to see if there are any hacks around this for now. Will report here if we find anything. If you find anything, let me know!

**Comment by markandrus** [comment]
In my experience testing with the Quickstart application, I only observed the crash when `detach`-ing the RemoteParticipants' RemoteAudioTrack/RemoteVideoTrack from their &lt;video&gt; element. There might exist a workaround that `detach`-es in a different way.

**Comment by wesleykendall** [comment]
@markandrus thanks for the suggestion! I just commented out detaching tracks and that seems to avoid the error, so I will experiment with this more as a possible work around

**Comment by wesleykendall** [comment]
For anyone that sees this issue, I worked around it with the following hack:

```
// Detach the Tracks from the DOM.
function detachTracks(tracks) {
  let isSafari = /Safari/.test(navigator.userAgent) && /Apple Computer/.test(navigator.vendor);
  tracks.forEach(function(track) {
    if (isSafari) {
      // Avoid detaching the track and only remove the DOM element for Safari.
      // This avoids the issue that happens because of https://github.com/twilio/twilio-video.js/issues/294
      // We are still uncertain about what side effects this problem ultimately has if particants
      // connect and reconnect multiple times.
      track._attachments.forEach(function(element) {
        element.remove();
      });
    } else {
      track.detach().forEach(function(detachedElement) {
        detachedElement.remove();
      });
    }
  });
}
```

I tried a couple other things, such as pausing the track before detaching and other things, but nothing seems to avoid the webkit bug. Anyways, this doesn't appear to be an issue in twilio-video.js, so feel free to close it if you'd like

**Comment by yogaboy** [comment]
@wesleykendall this resolves the issue completely for me. Thank you. I will take a look myself at the underlying code in the Twilio framework and if I find something relevant will connect in with Webkit bug tracker to report potential issues to them. This should not have to be worked around.

**Comment by rafaalb** [comment]
worked for me @wesleykendall thanks!

**Comment by vasan-emids** [comment]
Worked for us too, thanks a heap @wesleykendall !!!
